tracetools/span.go

```diff
@@ -2,28 +2,105 @@ package tracetools
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
-	"github.com/opentracing/opentracing-go/log"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
 	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
 )
 
-// StartSpanFromContext will start a span from the given context with the given
-// operation name. It will also do some common/repeated setup on the span to keep
-// code a little more DRY.
-func StartSpanFromContext(ctx context.Context, operation string) (opentracing.Span, context.Context) {
-	span, ctx := opentracing.StartSpanFromContext(ctx, operation)
-	// Make the span available for analytics in Datadog
-	span.SetTag(ddext.AnalyticsEvent, true)
-	return span, ctx
+// StartSpanFromContext will start a span appropriate to the given tracing backend from the given context with the given
+// operation name. It will also do some common/repeated setup on the span to keep code a little more DRY.
+// Will panic if tracerType is anything other than "datadog", "otel", or ""
+func StartSpanFromContext(ctx context.Context, operation string, tracerType string) (any, context.Context) {
+	switch tracerType {
+	case "datadog":
+		span, ctx := opentracing.StartSpanFromContext(ctx, operation)
+		span.SetTag(ddext.AnalyticsEvent, true) // Make the span available for analytics in Datadog
+		return span, ctx
+
+	case "otel":
+		ctx, span := otel.Tracer("buildkite_agent").Start(ctx, operation)
+		span.SetAttributes(attribute.String("analytics.event", "true"))
+		return span, ctx
+
+	case "":
+		return nil, ctx
+
+	default:
+		panic(fmt.Sprintf("Invalid tracing backend %q passed to tracetools.StartSpanFromContext", tracerType))
+	}
+}
+
+// Adds context to the span, either as opentracing tags, or as opentelemetry attributes
+// Will panic if span is anything other than an opentracing.Span, a trace.Span, or nil
+// Noop when span is nil
+func AddAttributesToSpan(span any, attributes map[string]string) {
+	switch span := span.(type) {
```

Comment:
most of these comments are to myself, but this one isn't: if anyone has any better ideas than "do a typeswitch every time you need to do something with a span" then i'm all ears, please come at me with better ideas than this

Reply:
>  if anyone has any better ideas than "do a typeswitch every time you need to do something with a span" then i'm all ears

Have your own Span interface with implementations for OTel or OpenTracing.

```go
type Span interface {
  AddAttributes(map[string]string)
  FinishWithError(error)
  RecordError(error)
}

func StartSpanFromContext(…) (Span, context.Context) { … }
```

Reply:
yeah, that is the obvious solution in retrospect 😅 just a smidge embarassing that yesterday-ben didn't come up with that

Reply:
i'll adjust this to use interface funtimes. i have a kind of halway solution locallty at the moment, i don't think it'll be too hard to adapt

Reply:
alright, that's donezo, way nicer now. not even a single panic! thanks for the help @cbandy ^_^